pkg/util/version/stream_test.go

```diff
@@ -40,52 +38,85 @@ func TestGetUpgradeStream(t *testing.T) {
+			upgradeY: false,
+			streams:  []*Stream{&stream43, &stream44},
+			want:     &stream43,
+		},
+		{
+			name:     "don't upgrade when Y versions match but current Z (4.3.19) is greater",
+			v:        NewVersion(4, 3, 19),
+			upgradeY: false,
+			streams:  []*Stream{&stream43, &stream44},
+			want:     nil,
 		},
 		{
-			name: "no upgrade when x.Y is higher than exected",
-			v:    NewVersion(4, 3, 19),
-			err:  fmt.Errorf("not upgrading: cvo desired version is 4.3.19"),
+			name:     "upgrade when Y versions match and candidate Z (4.4.3) is greater",
+			v:        NewVersion(4, 4, 2),
+			upgradeY: false,
+			streams:  []*Stream{&stream43, &stream44},
+			want:     &stream44,
 		},
 		{
-			name: " when X.y id lower than exected",
-			v:    NewVersion(4, 4, 2),
-			want: stream44,
+			name:     "don't upgrade when Y versions match but current Z (4.4.9) is greater",
+			v:        NewVersion(4, 4, 9),
+			upgradeY: false,
+			streams:  []*Stream{&stream43, &stream44},
+			want:     nil,
 		},
 		{
-			name: "no upgrade when X.y is higher than expected",
-			v:    NewVersion(4, 4, 9),
-			err:  fmt.Errorf("not upgrading: cvo desired version is 4.4.9"),
+			name:     "upgrade to Y+1 when allowed and candidate y.Z (4.3.18) < current y.Z (4.3.19)",
+			v:        NewVersion(4, 3, 19),
+			upgradeY: true,
+			streams:  []*Stream{&stream43, &stream44},
+			want:     &stream44,
 		},
 		{
-			name: "cvo error",
-			v:    NewVersion(4, 5, 1),
-			err:  fmt.Errorf("not upgrading: stream not found 4.5.1"),
+			name:     "upgrade to Y+1 when allowed and candidate y.Z == current y.Z (4.3.18)",
+			v:        stream43.Version,
+			upgradeY: true,
+			streams:  []*Stream{&stream43, &stream44},
+			want:     &stream44,
 		},
 		{
-			name: "error",
-			v:    NewVersion(5, 5, 1),
-			err:  fmt.Errorf("not upgrading: stream not found 5.5.1"),
+			name:     "upgrade to Y+1 (not Y+2) when allowed and candidate y.Z == current y.Z (4.3.18)",
+			v:        stream43.Version,
+			upgradeY: true,
+			streams:  []*Stream{&stream43, &stream44, &stream45},
+			want:     &stream44,
+		},
+		{
+			name:     "don't upgrade Y when not allowed",
+			v:        stream43.Version,
+			upgradeY: false,
+			streams:  []*Stream{&stream43, &stream44},
+			want:     nil,
 		},
 	} {
 		t.Run(tt.name, func(t *testing.T) {
-			Streams = []Stream{stream43, stream44}
-			got, err := GetUpgradeStream(tt.v)
-			if err != nil && tt.err != nil && !reflect.DeepEqual(tt.err, err) {
-				t.Fatal(err)
+			Streams = tt.streams
+			got := GetUpgradeStream(tt.v, tt.upgradeY)
+			if got != nil && tt.want != nil && !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("wanted %s, got %s", tt.want.Version, got.Version)
+			}
+			if got == nil && tt.want != nil {
+				t.Errorf("wanted %s, got nil", tt.want.Version)
 			}
-			if got != nil && !reflect.DeepEqual(got, &tt.want) {
-				t.Error(cmp.Diff(got, &tt.want))
+			if got != nil && tt.want == nil {
+				t.Errorf("wanted nil, got %s", got)
 			}
```

Comment:
overkill: reflect.DeepEqual is fine with nil input; cmp.Diff will print your diff.